### PR TITLE
handle rename failures similarly to s3a

### LINKED ISSUE #2059: rename: handle rename failures similarly to s3a
The goal of this issue is to modify LakefsFS rename operation to handle rename failure in the same manner s3a handles each case. 

s3a rename split into two parts: 
1. rename initialization: includes [input validation](https://github.com/apache/hadoop/blob/2960d83c255a00a549f8809882cd3b73a6266b6d/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java#L1609)
2. [rename execution](https://github.com/apache/hadoop/blob/2960d83c255a00a549f8809882cd3b73a6266b6d/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java#L1620) 

s3a handles failure in phase 1 by returning false, while failures in phase 2 trigger IOException.

### PR DESCRIPTION
**Pull request by talSofer**
Closes #2059

### CODE REVIEW

File: clients/hadoopfs/src/main/java/io/lakefs/LakeFSFileSystem.java [hunk 1]
```diff
@@ -263,6 +264,15 @@ public FSDataOutputStream append(Path path, int i, Progressable progressable) th
     public boolean rename(Path src, Path dst) throws IOException {
         ObjectLocation srcObjectLoc = pathToObjectLocation(src);
         ObjectLocation dstObjectLoc = pathToObjectLocation(dst);
+        if (srcObjectLoc.getPath().isEmpty()) {
+            LOG.error("rename: src {} is root directory", src);
```
Comment (arielshaqed) [inline clients/hadoopfs/src/main/java/io/lakefs/LakeFSFileSystem.java:268]: S3A does this?!  Weird, it looks as legitimate a rename as any.

File: clients/hadoopfs/src/main/java/io/lakefs/LakeFSFileSystem.java [hunk 2]
```diff
@@ -263,6 +264,15 @@ public FSDataOutputStream append(Path path, int i, Progressable progressable) th
...
+            return false;
+        }
+        if (dstObjectLoc.getPath().isEmpty()) {
+            LOG.error("rename: dst {} is root directory", dst);
```
Comment (arielshaqed) [inline clients/hadoopfs/src/main/java/io/lakefs/LakeFSFileSystem.java:272]: And the same here: S3A doesn't allow me to move all files under `s3a://bucket/oops` to be under `s3a://bucket/`?

File: clients/hadoopfs/src/main/java/io/lakefs/LakeFSFileSystem.java [hunk 3]
```diff
@@ -398,25 +406,37 @@ private boolean renameObject(LakeFSFileStatus srcStatus, Path dst) throws IOExce
...
+        switch (code){
+            case HttpStatus.SC_NOT_FOUND:
+                return new FileNotFoundException(msg);
+            case HttpStatus.SC_FORBIDDEN:
+                return new AccessDeniedException(msg);
+            default:
+                return new IOException(msg);
+        }
+    }
+
```
Comment (arielshaqed) [inline clients/hadoopfs/src/main/java/io/lakefs/LakeFSFileSystem.java:439]: Great stuff!

**Review by arielshaqed — COMMENTED** [review]
Thanks!  On the fence about approving because of the strange root directory behaviour.

**Review by arielshaqed — APPROVED** [review]
After verifying f2f that the weirdnesses are indeed S3A compatible... please just add links to this matching S3A behaviour.  I'm obviously (very) happy with the code.

Thanks!

File: clients/hadoopfs/src/main/java/io/lakefs/LakeFSFileSystem.java (hunk 1 above)
Comment (talSofer) [reply clients/hadoopfs/src/main/java/io/lakefs/LakeFSFileSystem.java:268]: yes, https://github.com/apache/hadoop/blob/2960d83c255a00a549f8809882cd3b73a6266b6d/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java#L1498

File: clients/hadoopfs/src/main/java/io/lakefs/LakeFSFileSystem.java (hunk 2 above)
Comment (talSofer) [reply clients/hadoopfs/src/main/java/io/lakefs/LakeFSFileSystem.java:272]: As discussed, s3a does not allow renaming a directory when the destination is a non-empty directory. I'm planning to handle this general case at https://github.com/treeverse/lakeFS/issues/2061. 

This line mimics what s3a does in https://github.com/apache/hadoop/blob/2960d83c255a00a549f8809882cd3b73a6266b6d/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java#L1501

File: clients/hadoopfs/src/main/java/io/lakefs/LakeFSFileSystem.java (hunk 3 above)
Comment (talSofer) [reply clients/hadoopfs/src/main/java/io/lakefs/LakeFSFileSystem.java:439]: Thanks!

**Review by nopcoder — APPROVED** [review]

**Pull request merged**